src/ocl/convolutionocl.cpp

```diff
@@ -655,49 +674,58 @@ static void DirConvFindCore(Handle& handle,
-        const auto algorithm_name = AlgorithmName{"miopenConvolutionFwdAlgoDirect"};
-        EvaluateInvokers(handle, all, algorithm_name, network_config, invoke_ctx, record);
+        if(!use_winograd_only)
+        {
+            ConvolutionUserBuffers bufs(workSpace, workSpaceSize);
+            bufs.SetFwd(x, w, y);
+            const auto direct_solutions = conv.FindDataDirectSolutions(
+                handle, xDesc, wDesc, yDesc, exhaustiveSearch, true, bufs, invoke_ctx);
+            const auto implictgemm_solutions = conv.FindDataImplicitGemmSolutions(
+                handle, xDesc, wDesc, yDesc, exhaustiveSearch, true, bufs, invoke_ctx);
+            const auto fft_solutions = conv.FindFftSolutions(ctx, invoke_ctx);
+            all_solutions.insert(
+                all_solutions.end(), direct_solutions.begin(), direct_solutions.end());
+            all_solutions.insert(
+                all_solutions.end(), implictgemm_solutions.begin(), implictgemm_solutions.end());
+            all_solutions.insert(all_solutions.end(), fft_solutions.begin(), fft_solutions.end());
+        }
     }
 
-    // Implicit GEMM algo
-    if(!use_winograd_only)
+    // Precompile Solutions
     {
-        ConvolutionUserBuffers bufs(workSpace, workSpaceSize);
-        bufs.SetFwd(x, w, y);
-        const auto all = conv.FindDataImplicitGemmSolutions(
-            handle, xDesc, wDesc, yDesc, exhaustiveSearch, true, bufs, invoke_ctx);
-        PrecompileSolutions(handle, all);
-        const auto algorithm_name = AlgorithmName{"miopenConvolutionFwdAlgoImplicitGEMM"};
-        EvaluateInvokers(handle, all, algorithm_name, network_config, invoke_ctx, record);
+        PrecompileSolutions(handle, all_solutions);
     }
 
-    // FFT algo
-    if(!use_winograd_only && !miopen::IsDisabled(MIOPEN_DEBUG_CONV_FFT{}))
+    // Evaluate Invokers
+    auto iter_start = all_solutions.begin();
+    auto iter       = all_solutions.begin();
+    std::string algo_name =
+        solver::Id(iter_start->solver_id).GetAlgo(ctx.conv_problem.GetDirection());
+    for(; iter != all_solutions.end(); ++iter)
     {
-        const auto all            = FindAllFFTSolutions(ctx, invoke_ctx);
-        const auto algorithm_name = AlgorithmName{"miopenConvolutionFwdAlgoFFT"};
-        PrecompileSolutions(handle, all);
-        EvaluateInvokers(handle, all, algorithm_name, network_config, invoke_ctx, record);
+        if(solver::Id(iter->solver_id).GetAlgo(ctx.conv_problem.GetDirection()) != algo_name)
+        {
+            EvaluateInvokers(handle,
+                             {iter_start, iter},
+                             AlgorithmName(algo_name),
+                             network_config,
+                             invoke_ctx,
+                             record);
+            iter_start = iter;
+            algo_name  = solver::Id(iter_start->solver_id).GetAlgo(ctx.conv_problem.GetDirection());
+        }
     }
+    EvaluateInvokers(
```

Comment:
Quiet change. Current develop: the library evaluates available solutions per algorithm and adds a "best" record per each. The result is up to 5 (GEMM, Direct, FFT, Winograd, iGemm) best kernels. The caller then can select the one that suits its needs (e.g. it can skip the fastest GEMM because it needs more workspace then available and take the next Direct, or it can skip the fastest Winograd because its numeric traits are worse than Direct or GEMM).

This PR: returns only the fastest solution.

This behavior change should be addressed.

Reply:
```
        //Evaluate Invokers
        auto iter_start = all_solutions.begin();
        auto iter       = all_solutions.begin();
        std::string algo_name =
            solver::Id(iter_start->solver_id).GetAlgo(ctx.conv_problem.GetDirection());
        for(; iter != all_solutions.end(); ++iter)
        {
            if(solver::Id(iter->solver_id).GetAlgo(ctx.conv_problem.GetDirection()) != algo_name)
            {
                EvaluateInvokers(handle,
                                 {iter_start, iter},
                                 AlgorithmName(algo_name),
                                 network_config,
                                 invoke_ctx,
                                 record);
                iter_start = iter;
                algo_name =
                    solver::Id(iter_start->solver_id).GetAlgo(ctx.conv_problem.GetDirection());
            }
        }
        EvaluateInvokers(handle,
                         {iter_start, iter},
                         AlgorithmName(algo_name),
                         network_config,
                         invoke_ctx,
                         record);
```
In the evalution process, I colloect all solutions that use same algorithm togther and call `EvaluateInvokers`. This is the same process as current develop and also adds a "best" record per algorithm. The `EvaluateInvokers` called outside the for loop is to ensure that the last segment(`iter_start` to `all_solutions.end()`) of `all_solutions` will be evaluated rather then abandoned.

Reply:
My initial review comment was invalid. I seem to have lost my focus :face_with_head_bandage: 

Resolved.